src/mlpack/tests/ne_test.cpp

```diff
@@ -0,0 +1,167 @@
+  BOOST_REQUIRE_EQUAL(linkGene.Weight(), 10);
+
+  // Test set function.
+  linkGene.Weight(20);
+  BOOST_REQUIRE_EQUAL(linkGene.Weight(), 20);
+
+  // Test copy constructor.
+  LinkGene linkGene2(linkGene);
+  BOOST_REQUIRE_EQUAL(linkGene2.FromNeuronId(), 1);
+
+  // Test operator =.
+  LinkGene linkGene3(1, 3, 1, 7);
+  linkGene = linkGene3;
+  BOOST_REQUIRE_EQUAL(linkGene3.InnovationId(), 1);
+
+}
+
+/**
+ * Test NeuronGene class.
+ */
+BOOST_AUTO_TEST_CASE(NENeuronGeneTest)
+{
+  // Create a neuron gene.
+  ActivationFuncType actFuncType = SIGMOID;
+  NeuronType neuronType = INPUT;
+  NeuronGene neuronGene(1, neuronType, actFuncType, 0, 0);
+
+  // Test parametric constructor and access functions.
+  BOOST_REQUIRE_EQUAL(neuronGene.Id(), 1);
+  BOOST_REQUIRE_EQUAL(neuronGene.Type(), INPUT);
+  BOOST_REQUIRE_EQUAL(neuronGene.ActFuncType(), SIGMOID);
+  BOOST_REQUIRE_EQUAL(neuronGene.aInput, 0);
+  BOOST_REQUIRE_EQUAL(neuronGene.aActivation, 0);
+
+  // Test set function
+  neuronGene.ActFuncType(RELU);
+  BOOST_REQUIRE_EQUAL(neuronGene.ActFuncType(), RELU);
+
+  // Test copy function.
+  NeuronGene neuronGene2(neuronGene);
+  BOOST_REQUIRE_EQUAL(neuronGene2.Id(), 1);
+
+  // Test operator =.
+  NeuronGene neuronGene3(11, neuronType, actFuncType, 0, 0);
+  neuronGene = neuronGene3;
+  BOOST_REQUIRE_EQUAL(neuronGene.Id(), 11);
+
+}
+
+/**
+ * Test Genome class.
+ */
+BOOST_AUTO_TEST_CASE(NEGenomeTest)
+{
+
+}
+
+/**
+ * Test Population class.
+ */
+BOOST_AUTO_TEST_CASE(NEPopulationTest)
+{
+
+}
+
+/**
+ * Test CNE by XOR task.
+ */
+BOOST_AUTO_TEST_CASE(NECneXorTest)
+{
+  mlpack::math::RandomSeed(1);
+
+  // Set CNE algorithm parameters.
+  Parameters params;
+  params.aPopulationSize = 500;
+  params.aMutateRate = 0.2;
+  params.aCrossoverRate = 0.3;
+
+  // Construct seed genome for xor task.
+  size_t id = 0;
+  size_t numInput = 3;
+  size_t numOutput = 1;
+  size_t depth = 2;
+  double fitness = -1;
+  std::vector<NeuronGene> neuronGenes;
+  std::vector<LinkGene> linkGenes;
+
+  NeuronGene inputGene1(0, INPUT, SIGMOID, 0, 0);
+  NeuronGene inputGene2(1, INPUT, SIGMOID, 0, 0);
+  NeuronGene biasGene(2, BIAS, LINEAR, 0, 0);
+  NeuronGene outputGene(3, OUTPUT, SIGMOID, 0, 0);
+  NeuronGene hiddenGene(4, HIDDEN, SIGMOID, 0, 0);
+  
+  neuronGenes.push_back(inputGene1);
+  neuronGenes.push_back(inputGene2);
+  neuronGenes.push_back(biasGene);
+  neuronGenes.push_back(outputGene);
+  neuronGenes.push_back(hiddenGene);
+
+  LinkGene link1(0, 3, 0, 0);
+  LinkGene link2(1, 3, 0, 0);
+  LinkGene link3(2, 3, 0, 0);
+  LinkGene link4(0, 4, 0, 0);
+  LinkGene link5(1, 4, 0, 0);
+  LinkGene link6(2, 4, 0, 0);
+  LinkGene link7(4, 3, 0, 0);
+
+  linkGenes.push_back(link1);
+  linkGenes.push_back(link2);
+  linkGenes.push_back(link3);
+  linkGenes.push_back(link4);
+  linkGenes.push_back(link5);
+  linkGenes.push_back(link6);
+  linkGenes.push_back(link7);
+
+  Genome seedGenome = Genome(0, neuronGenes, linkGenes, numInput, numOutput, depth, fitness);
+
```

Comment:
I think it would be a good idea to set a default initial genome inside the CNE method. The authors start with a genome where each input unit is connected with the output units. But, we could also use a hidden units as you did. This way, we can directly start by evolving the network.